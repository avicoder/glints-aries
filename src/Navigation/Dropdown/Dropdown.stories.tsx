import React from 'react';
import { Story, Meta } from '@storybook/react';

import { Dropdown, Props } from './Dropdown';
import {
  ArrowNextIcon,
  ArrowNextDoubleIcon,
  InfoIcon,
} from '../../General/Icon/components';

export default {
  title: 'Navigation/Dropdown',
  component: Dropdown,
  argTypes: {
    iconDefaultColor: {
      control: 'color',
    },
    leftIcon: {
      control: {
        type: 'select',
        options: {
          'Example: ArrowBackIcon': <ArrowNextIcon />,
          'Example: ArrowNextDoubleIcon': <ArrowNextDoubleIcon />,
        },
      },
    },
  },
} as Meta;

const Template: Story<Props> = args => {
  return (
    <div style={{ height: '150px' }}>
      <Dropdown label="Career" {...args}>
        <Dropdown.Item value="pm">Product Manager</Dropdown.Item>
        <Dropdown.Item value="se">Software Engineer</Dropdown.Item>
      </Dropdown>
    </div>
  );
};

export const Default = Template.bind({});

export const HoverToOpen = Template.bind({});
HoverToOpen.args = {
  hoverToOpen: true,
};

export const DropdownWithIconAndDefaultIconColour = Template.bind({});
DropdownWithIconAndDefaultIconColour.args = {
  leftIcon: <InfoIcon />,
  iconDefaultColor: 'red',
};

export const DropdownWithHoverLineEffect = Template.bind({});
DropdownWithHoverLineEffect.args = {
  showHoverLine: true,
};

export const DropdownWithRightPlacement = Template.bind({});
DropdownWithRightPlacement.args = {
  dropDownPlacement: 'right',
};

export const DropdownWithFullWidth = Template.bind({});
DropdownWithFullWidth.args = {
  showFullWidth: true,
};

export const DisabledDropdown = Template.bind({});
DisabledDropdown.args = {
  disabled: true,
};
