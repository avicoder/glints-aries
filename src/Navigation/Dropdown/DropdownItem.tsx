import * as React from 'react';

import { DropdownItemWrapper } from './DropdownStyle';

const DropdownItem: React.FunctionComponent<Props> = props => {
  const { children, className, ...defaultProps } = props;

  return (
    <DropdownItemWrapper className={className} {...defaultProps}>
      {children}
    </DropdownItemWrapper>
  );
};

export interface Props
  extends React.ComponentPropsWithoutRef<typeof DropdownItemWrapper> {
  value?: any;
  onClick?: (e: React.MouseEvent<HTMLLIElement, MouseEvent>) => void;
  children: React.ReactNode;
}

export default DropdownItem;
