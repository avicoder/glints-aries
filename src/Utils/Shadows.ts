export const Shadows = {
  1: {
    up: '0 2px 4px 0 rgba(0, 0, 0, 0.15)',
  },
  2: {
    up: '0 4px 8px 0 rgba(0, 0, 0, 0.15)',
  },
  3: {
    up: '0 6px 12px 0 rgba(0, 0, 0, 0.10)',
  },
  4: {
    up: '0 12px 24px 0 rgba(0, 0, 0, 0.10)',
    down: '0 -6px 12px 0 rgba(0, 0, 0, 0.10)',
  },
};
