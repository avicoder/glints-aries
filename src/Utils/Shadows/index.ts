const ShadowOneDown = '0 2px 4px 0 rgba(0, 0, 0, 0.15)';
const ShadowTwoDown = '0 4px 8px 0 rgba(0, 0, 0, 0.15)';
const ShadowThreeDown = '0 6px 12px 0 rgba(0, 0, 0, 0.10)';
const ShadowFiveDown = '0 12px 24px 0 rgba(0, 0, 0, 0.10)';
const ShadowFiveUp = '0 -6px 12px 0 rgba(0, 0, 0, 0.10)';

export const Shadows = {
  1: {
    down: ShadowOneDown,
  },
  2: {
    down: ShadowTwoDown,
  },
  3: {
    down: ShadowThreeDown,
  },
  4: {
    down: ShadowFiveDown,
    up: ShadowFiveUp,
  },
} as const;

export type ShadowValueType =
  | typeof ShadowOneDown
  | typeof ShadowTwoDown
  | typeof ShadowThreeDown
  | typeof ShadowFiveDown
  | typeof ShadowFiveUp;
