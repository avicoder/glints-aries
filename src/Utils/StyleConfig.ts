export enum ETabAlignment {
  HORIZONTAL = 'horizontal',
  VERTICAL = 'vertical',
}

export enum ETabThemeVariant {
  BLUE = 'blue',
  BLACK = 'black',
}

export enum ETooltipPosition {
  TOP = 'top',
  BOTTOM = 'bottom',
  LEFT = 'left',
  RIGHT = 'right',
}

export const ButtonVariant = {
  SOLID_WHITE: 'solid-white',
  SOLID_BLUE: 'solid-blue',
  YELLOW: 'yellow',
  GHOST: 'ghost',
  WHITE_GREY: 'white-grey',
  LINK: 'link',
};

export const RoundedCornerButtonVariant = {
  SOLID_WHITE: 'solid-white',
  SOLID_BLUE: 'solid-blue',
  GHOST: 'ghost',
  WHITE_GREY: 'white-grey',
};

export const PsychedelicTheme = {
  BLUE_DOMINANT: 'blue-dominant',
  BLUE_DEFAULT: 'blue-default',
  RED_DOMINANT: 'red-dominant',
  RED_DEFAULT: 'red-default',
  YELLOW_DOMINANT: 'yellow-dominant',
  YELLOW_DEFAULT: 'yellow-default',
};

export const ButtonTheme = {
  BLUE: 'blue',
  WHITE: 'white',
};

export const ScreenSize = {
  mobileS: 320,
  mobileM: 480,
  mobileL: 640,
  tablet: 768,
  desktopS: 1024,
  desktopM: 1260,
  desktopL: 1440,
};

export const Device = {
  mobileS: `(max-width: ${ScreenSize.mobileS}px)`,
  mobileM: `(max-width: ${ScreenSize.mobileM}px)`,
  mobileL: `(max-width: ${ScreenSize.mobileL}px)`,
  tablet: `(max-width: ${ScreenSize.tablet}px)`,
  desktopS: `(max-width: ${ScreenSize.desktopS}px)`,
  desktopM: `(max-width: ${ScreenSize.desktopM}px)`,
  desktopL: `(max-width: ${ScreenSize.desktopL}px)`,
  desktopXL: `(min-width: ${ScreenSize.desktopL + 1}px)`,
};
