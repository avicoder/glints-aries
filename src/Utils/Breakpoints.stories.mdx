import { Meta } from '@storybook/addon-docs/blocks';
import { ScreenSize, Device } from './StyleConfig';

<Meta title="Utilities/Breakpoints" />

# Breakpoints

These utilities are used internally by components like
[GlintsContainer](/?path=/docs/utilities-glintscontainer--page#glintscontainer).
You should use them when building your own responsive styles to ensure that
breaks only happen at a few well-defined viewport-widths.

## Device

Our device classes. Names are generally oriented upwards: For example, when you
use `Device.desktopS`, your styles will apply for all viewport-widths greater or
equal to `1024px`. That means that you have to override those styles with rules
for wider viewports if needed. Thus, if you have styles for serveral widths, the
rules for narrow viewports should be at the top and should be consecutively
overridden by rules for wider viewports towards the bottom of your css.

`mobileS` is an exception: It applies "downwards", i.e. on all screens smaller
than `480px`.

```TypeScript
import { Device } from 'glints-aries';

const MyContainer = styled.div`
  // Applies on screens smaller than tablet
  color: red;

  // Applies from tablets to screens smaller than large desktops
  @media ${Device.tablet} {
    color: green;
  }

  // Applies on large desktops
  @media ${Device.desktopL} {
    color: blue;
  }
`;
```

<table>
  <thead>
    <tr>
      <th>Device</th>
      <th>Value</th>
    </tr>
  </thead>
  <tbody>
    {Object.keys(Device).map(dev => (
      <>
        <tr>
          <td>{dev}</td>
          <td>{Device[dev]}</td>
        </tr>
      </>
    ))}
  </tbody>
</table>

## ScreenSize

These are the numeric pixel values that define our breakpoints. The values are
the smallest pixel width that should be classified as the respective constant's
size class. For example, a viewport needs to be at least `1440px` wide to "qualify" as
Desktop Large. Note that the values do not include the `px` prefix;

```TypeScript
import { ScreenSize } from 'glints-aries';

const MyContainer = styled.div`
  color: red;
  @media (min-width: ${ScreenSize.desktopL}px) {
    color: blue;
  }
`;
```

<table>
  <thead>
    <tr>
      <th>Device</th>
      <th>Value</th>
    </tr>
  </thead>
  <tbody>
    {Object.keys(ScreenSize).map(size => (
      <>
        <tr>
          <td>{size}</td>
          <td>{ScreenSize[size]}</td>
        </tr>
      </>
    ))}
  </tbody>
</table>
