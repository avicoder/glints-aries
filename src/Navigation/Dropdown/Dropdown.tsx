import * as React from 'react';

import classNames from 'classnames';

import { ArrowDownIcon } from '../../General/Icon/components';

import { escEvent } from '../../Utils/DomUtils';

import {
  DropdownContainer,
  DropdownWrapper,
  DropdownHeader,
  DropdownBody,
  DropdownItemWrapper,
  IconWrapper,
} from './DropdownStyle';

import DropdownItem, { Props as DropdownItemProps } from './DropdownItem';

export const Dropdown = ({
  label,
  showHoverLine = false,
  dropDownPlacement = 'left',
  noLineBreak = false,
  iconDefaultColor = 'black',
  showFullWidth = false,
  disabled = false,
  hoverToOpen = false,
  onChange,
  children,
  leftIcon,
  className,
  itemElement,
  ...defaultProps
}: Props) => {
  const dropdownBodyRef = React.useRef<HTMLUListElement>(null);

  const [dropdownLabel, setDropdownLabel] = React.useState<string>(label);
  const [isOpen, setIsOpen] = React.useState<boolean>(false);
  const [cursor, setCursor] = React.useState<number>(0);

  const handleOpen = () => {
    if (!disabled) {
      setIsOpen(!isOpen);
    }
  };

  const hoverOpen = () => {
    if (hoverToOpen) {
      setIsOpen(true);
    }
  };

  const hoverClose = () => {
    if (hoverToOpen) {
      setIsOpen(false);
      setCursor(0);
    }
  };

  const handleClose = () => {
    setIsOpen(false);
    setCursor(0);
  };

  const handleClickItem = (
    onClick: (e: React.MouseEvent<HTMLLIElement, MouseEvent>) => void
  ) => (e: React.MouseEvent<HTMLLIElement, MouseEvent>) => {
    const itemElement = e.target as HTMLLIElement;

    if (itemElement.dataset.value) {
      setDropdownLabel(itemElement.innerHTML);
      setIsOpen(false);

      if (onChange !== undefined) {
        onChange(itemElement.dataset.value);
      }
    } else {
      setIsOpen(false);
    }

    if (onClick !== undefined) {
      onClick(e);
    }
  };

  const handleMouseEnter = setCursor;

  const handleKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
    e.preventDefault();

    // Up Arrow
    if (e.keyCode === 38 && cursor > 0) {
      setCursor(cursor - 1);
    } else if (
      e.keyCode === 40 && // Down Arrow
      cursor < React.Children.count(children) - 1
    ) {
      setCursor(cursor + 1);
    } else if (e.keyCode === 13) {
      // Enter
      setDropdownLabel(
        dropdownBodyRef.current.querySelector('.active').innerHTML
      );
      setIsOpen(false);
    }
  };

  React.useEffect(() => {
    const handleEscKeydown = escEvent(handleClose);
    document.addEventListener('keydown', handleEscKeydown, false);
    return () => {
      document.removeEventListener('keydown', handleEscKeydown, false);
    };
  }, []);

  const LeftIcon = leftIcon;

  return (
    <DropdownContainer
      className={classNames('aries-dropdown', className)}
      tabIndex={0}
      onClick={handleOpen}
      onMouseEnter={hoverOpen}
      onMouseLeave={hoverClose}
      onBlur={handleClose}
      onKeyDown={handleKeyDown}
      role="menuitem"
      aria-label={dropdownLabel}
      aria-haspopup="true"
      {...defaultProps}
    >
      <DropdownWrapper className="dropdown-contentwrapper" tabIndex={-1}>
        <DropdownHeader
          className="dropdown-content"
          isOpen={isOpen}
          disabled={disabled}
          showHoverLine={showHoverLine}
          showFullWidth={showFullWidth}
          iconDefaultColor={iconDefaultColor}
        >
          {LeftIcon}
          <span>{itemElement ? itemElement : dropdownLabel}</span>
          <IconWrapper isOpen={isOpen}>
            <ArrowDownIcon color={!disabled ? iconDefaultColor : '#777777'} />
          </IconWrapper>
        </DropdownHeader>
        <DropdownBody
          className="dropdown-listbox"
          role="listbox"
          aria-hidden={!isOpen && true}
          onClick={(e: React.MouseEvent) => e.stopPropagation()}
          open={isOpen}
          dropDownPlacement={dropDownPlacement}
          noLineBreak={noLineBreak}
          showFullWidth={showFullWidth}
          showHoverLine={showHoverLine}
          ref={dropdownBodyRef}
        >
          {React.Children.toArray(children)
            .filter(Boolean)
            .map((item: React.ReactElement<DropdownItemProps>, index) => {
              const dropDownItemClassName = item.props.className || '';
              return (
                <DropdownItemWrapper
                  className={classNames(
                    { active: cursor === index },
                    `${dropDownItemClassName}`
                  )}
                  role="option"
                  data-value={item.props.value}
                  key={item.key}
                  onMouseDown={handleClickItem(item.props.onClick)}
                  onMouseEnter={() => handleMouseEnter(index)}
                  tabIndex={0}
                  showFullWidth={showFullWidth}
                >
                  {item.props.children}
                </DropdownItemWrapper>
              );
            })}
        </DropdownBody>
      </DropdownWrapper>
    </DropdownContainer>
  );
};

Dropdown.Item = DropdownItem;

type PropsFromDropdownContainer = Omit<
  React.ComponentPropsWithoutRef<typeof DropdownContainer>,
  'onChange'
>;

type PropsFromDropdownHeader = Omit<
  React.ComponentPropsWithoutRef<typeof DropdownHeader>,
  'isOpen'
>;

type PropsFromDropdownBody = Omit<
  React.ComponentPropsWithoutRef<typeof DropdownBody>,
  'open'
>;

export interface Props
  extends PropsFromDropdownContainer,
    PropsFromDropdownHeader,
    PropsFromDropdownBody {
  children: React.ReactNode;
  /** Sets label of Dropdown. */
  label: string;
  /** Opens Dropdown on hovering. */
  hoverToOpen?: boolean;
  /** Icon to show at the left side of dropdown button. */
  leftIcon?: React.ReactNode;
  /**	Dropdown label in component. */
  itemElement?: React.ReactNode;
  /** Dropdown icon color */
  iconDefaultColor?: string;
  /** To get value of Dropdown's item. */
  onChange?(value: string): void;
  /** Show underline when hovered. */
  showHoverLine?: boolean;
  /** Placement position for dropdown. */
  dropDownPlacement?: 'left' | 'right';
  /**	No break line in dropdown. */
  noLineBreak?: boolean;
  /**	Dropdown take full width. */
  showFullWidth?: boolean;
  /** Disabling the dropdown. */
  disabled?: boolean;
}

export default Dropdown;
