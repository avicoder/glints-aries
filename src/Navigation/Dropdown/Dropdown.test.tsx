import * as React from 'react';

import * as renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect';
import {
  fireEvent,
  render,
  queryByText,
  queryByRole,
  queryByTestId,
} from '@testing-library/react';

import Dropdown from './Dropdown';
import DropdownItem from './DropdownItem';
import { Greyscale, SecondaryColor } from '../../Utils/Colors';

describe('<Dropdown/> render', () => {
  test('should match snapshot', () => {
    const snapshot = renderer
      .create(
        <Dropdown label="Career">
          <DropdownItem value="pm">Product Manager</DropdownItem>
          <DropdownItem value="se">Software Engineer</DropdownItem>
        </Dropdown>
      )
      .toJSON();
    expect(snapshot).toMatchSnapshot();
  });

  test('options should not be visible when it is rendered', () => {
    const { queryByText } = render(
      <Dropdown label="Career">
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );

    const optionPM = queryByText('Product Manager');
    const optionSE = queryByText('Software Engineer');

    expect(optionPM).not.toBeVisible();
    expect(optionSE).not.toBeVisible();
  });

  test('should be able to render with null children', () => {
    render(
      <Dropdown label="Career">
        {null}
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );

    const options = document.querySelector('.dropdown-listbox').children;
    expect(options.length).toEqual(2);
  });
});

describe('<Dropdown/> with DropdownHeader props', () => {
  // TODO: find a way to test showHoverLine, which needs pseudo element style validation
  test('disabled', () => {
    render(
      <Dropdown label="Career" disabled>
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );
    expect(document.querySelector('.dropdown-content')).toHaveStyle(
      'cursor: not-allowed'
    );
  });

  test('showFullWidth', () => {
    render(
      <Dropdown label="Career" showFullWidth>
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );
    expect(document.querySelector('.dropdown-content')).toHaveStyle(
      'width: 100%; padding: 10px 0 10px 20px;'
    );
  });
});

describe('<Dropdown/> with DropdownBody props', () => {
  test('dropDownPlacement', () => {
    render(
      <Dropdown label="Career" dropDownPlacement="right">
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );
    expect(document.querySelector('.dropdown-listbox')).toHaveStyle(
      'right: 0; width: auto;'
    );
  });

  test('noLineBreak', () => {
    render(
      <Dropdown label="Career" noLineBreak>
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );
    expect(document.querySelector('.dropdown-listbox')).toHaveStyle(
      'white-space: normal'
    );
  });

  test('showFullWidth', () => {
    render(
      <Dropdown label="Career" showFullWidth>
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );
    expect(document.querySelector('.dropdown-listbox')).toHaveStyle(
      'width: calc(100% - 20px); left: 10px;'
    );
  });

  test('showHoverLine', () => {
    render(
      <Dropdown label="Career" showHoverLine>
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );
    expect(document.querySelector('.dropdown-listbox')).toHaveStyle(
      'margin-top: 13px;'
    );
  });
});

describe('<Dropdown/> mouse event', () => {
  test('options should appear when clicking on it, if prop hoverToOpen is falsy', () => {
    const { queryByRole, queryByText } = render(
      <Dropdown label="Career">
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );

    const dropdown = queryByRole('menuitem');
    fireEvent.click(dropdown);

    const optionPM = queryByText('Product Manager');
    const optionSE = queryByText('Software Engineer');

    expect(optionPM).toBeVisible();
    expect(optionSE).toBeVisible();
  });

  test('options should appear when mouse hovering on it, if prop hoverToOpen is truthy', () => {
    const { queryByRole, queryByText } = render(
      <Dropdown label="Career" hoverToOpen={true}>
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );

    const dropdown = queryByRole('menuitem');
    fireEvent.mouseOver(dropdown);

    const optionPM = queryByText('Product Manager');
    const optionSE = queryByText('Software Engineer');

    expect(optionPM).toBeVisible();
    expect(optionSE).toBeVisible();
  });

  test('dropdown button should be highlighted when mouse is hovering on', () => {
    render(
      <Dropdown label="Career" hoverToOpen={true}>
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );

    const dropdownHeader = document.querySelector('.dropdown-content');
    fireEvent.mouseOver(dropdownHeader);

    expect(dropdownHeader).toHaveStyle(`background: ${Greyscale.softgrey}`);
  });

  test('option should be highlighted when mouse is hovering on', () => {
    const { queryByRole, queryByText } = render(
      <Dropdown label="Career" hoverToOpen={true}>
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );

    const dropdown = queryByRole('menuitem');
    fireEvent.click(dropdown);

    const optionPM = queryByText('Product Manager');
    const optionSE = queryByText('Software Engineer');
    fireEvent.mouseOver(optionPM);

    expect(optionPM).toHaveStyle(`background: ${Greyscale.softgrey}`);
    expect(optionSE).not.toHaveStyle(`background: ${Greyscale.softgrey}`);
  });

  test('selected option should be displayed', async () => {
    render(
      <Dropdown label="Career">
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );

    const dropdown = queryByRole(document.body, 'menuitem');
    fireEvent.click(dropdown);
    fireEvent.mouseDown(queryByText(document.body, 'Product Manager'));

    expect(document.querySelector('.dropdown-listbox')).not.toBeVisible();
    expect(
      queryByText(
        document.querySelector('.dropdown-content'),
        'Product Manager'
      )
    ).toBeVisible();
  });

  test('arrow icon should be rotated by different angles when dropdown is opened/closed', () => {
    render(
      <Dropdown label="Career">
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );

    const arrowIcon = queryByTestId(document.body, 'icon-svg');
    expect(arrowIcon).toHaveStyle('transform: rotate(0)');

    const dropdown = queryByRole(document.body, 'menuitem');
    fireEvent.click(dropdown);
    expect(arrowIcon).toHaveStyle('transform: rotate(180deg)');
  });
});

describe('<Dropdown/> keydown event', () => {
  test('press up arrow key and down arrow key should should change highlighted option', async () => {
    render(
      <Dropdown label="Career">
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );

    const dropdown = queryByRole(document.body, 'menuitem');
    fireEvent.click(dropdown);

    const optionPM = queryByText(document.body, 'Product Manager');
    const optionSE = queryByText(document.body, 'Software Engineer');
    expect(optionPM).toHaveStyle(`background: ${Greyscale.softgrey}`);
    expect(optionSE).not.toHaveStyle(`background: ${Greyscale.softgrey}`);

    fireEvent.keyDown(dropdown, { key: 'DownArrow', keyCode: 40 });
    expect(optionPM).not.toHaveStyle(`background: ${Greyscale.softgrey}`);
    expect(optionSE).toHaveStyle(`background: ${Greyscale.softgrey}`);

    fireEvent.keyDown(dropdown, { key: 'UpArrow', keyCode: 38 });
    expect(optionPM).toHaveStyle(`background: ${Greyscale.softgrey}`);
    expect(optionSE).not.toHaveStyle(`background: ${Greyscale.softgrey}`);
  });

  test('press enter key should select an highlighted option', async () => {
    render(
      <Dropdown label="Career">
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );

    const dropdown = queryByRole(document.body, 'menuitem');
    fireEvent.click(dropdown);

    const optionPM = queryByText(document.body, 'Product Manager');
    expect(optionPM).toHaveStyle(`background: ${Greyscale.softgrey}`);

    fireEvent.keyDown(dropdown, { key: 'Enter', keyCode: 13 });
    expect(optionPM).not.toBeVisible();

    expect(
      queryByText(
        document.querySelector('.dropdown-content'),
        'Product Manager'
      )
    ).toBeVisible();
  });

  test('press esc key should close dropdown', async () => {
    render(
      <Dropdown label="Career">
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );

    const dropdown = queryByRole(document.body, 'menuitem');
    fireEvent.click(dropdown);

    const dropdownContent = queryByRole(document.body, 'listbox');
    expect(dropdownContent).toBeVisible();

    fireEvent.keyDown(document, { key: 'Esc', keyCode: 27 });
    expect(dropdownContent).not.toBeVisible();
  });

  test('handleEscKeydown listener should be removed before component unmount', async () => {
    const removeEventListenerSpy = jest.spyOn(document, 'removeEventListener');
    const { unmount } = render(
      <Dropdown label="Career">
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );

    unmount();
    expect(removeEventListenerSpy).toHaveBeenCalledTimes(1);
    expect(
      (removeEventListenerSpy.mock.calls[0][1] as () => void).name
    ).toEqual('listener');
    removeEventListenerSpy.mockRestore();
  });
});

describe('<Dropdown/> logic', () => {
  test('onChange callback should be called with correct option value', async () => {
    const onChange = jest.fn();
    render(
      <Dropdown label="Career" onChange={onChange}>
        <DropdownItem value="pm">Product Manager</DropdownItem>
        <DropdownItem value="se">Software Engineer</DropdownItem>
      </Dropdown>
    );

    const dropdown = queryByRole(document.body, 'menuitem');
    fireEvent.click(dropdown);
    const option = queryByText(document.body, 'Product Manager');
    fireEvent.mouseDown(option);

    expect(onChange.mock.calls[0][0]).toEqual('pm');
  });
});
