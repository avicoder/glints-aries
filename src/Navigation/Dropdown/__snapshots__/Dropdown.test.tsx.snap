// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown/> render should match snapshot 1`] = `
<div
  aria-haspopup="true"
  aria-label="Career"
  className="DropdownStyle__DropdownContainer-sc-1sc3c6h-1 dneJaG aries-dropdown"
  onBlur={[Function]}
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="menuitem"
  tabIndex={0}
>
  <div
    className="DropdownStyle__DropdownWrapper-sc-1sc3c6h-0 iYCvaA dropdown-contentwrapper"
    tabIndex={-1}
  >
    <div
      className="DropdownStyle__DropdownHeader-sc-1sc3c6h-2 jTRVzJ dropdown-content"
      disabled={false}
    >
      <span>
        Career
      </span>
      <span
        className="DropdownStyle__IconWrapper-sc-1sc3c6h-5 jOTbED"
      >
        <svg
          className="IconStyle__VerticalCenteredSvg-x7d280-0 cJSNzi"
          data-testid="icon-svg"
          fill="black"
          height="1em"
          viewBox="0 0 100 100"
          width="1em"
        >
          <polygon
            points="11.7984725 19 49.9935275 57.3950291 88.1885825 19 99.987055 30.7984725 49.9935275 80.792 0 30.7984725"
          />
        </svg>
      </span>
    </div>
    <ul
      aria-hidden={true}
      className="DropdownStyle__DropdownBody-sc-1sc3c6h-3 ustjj dropdown-listbox"
      onClick={[Function]}
      open={false}
      role="listbox"
    >
      <li
        className="DropdownStyle__DropdownItemWrapper-sc-1sc3c6h-4 bnpgGr active"
        data-value="pm"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        role="option"
        tabIndex={0}
      >
        Product Manager
      </li>
      <li
        className="DropdownStyle__DropdownItemWrapper-sc-1sc3c6h-4 bnpgGr"
        data-value="se"
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        role="option"
        tabIndex={0}
      >
        Software Engineer
      </li>
    </ul>
  </div>
</div>
`;
